MILESTONE-1.md

```diff
@@ -0,0 +1,101 @@
+# Milestone 1
+
+The first milestone includes 3 contracts (in red on the diagram below): A space contract holding votes and proposals, an authenticator contract approving signed messages for casting a vote or creating a proposal, and a voting strategy where 1 address is equal to 1 voting power. Execution, custom settings, or storage proof calculations are not part of this milestone.
+
+![](https://user-images.githubusercontent.com/16245250/155883647-8f853411-bea9-4a73-bd69-2be850d4ffc9.png)
+
+Source: https://whimsical.com/snapshot-x-design-proposal-Gy1y3P6zTs4rdE4cStaKW4
+
+## Requirements
+
+Here are the actions to cover in this milestone:
+- Use auth contract as proxy
+- Create a proposal
+- Cast a vote
+- Store votes and proposals
+
+## Interfaces
+
+### Space contract
+
+This is the main contract, previously called "voting contract". The contract is responsible for storing proposals and votes, it also defines and stores parameters for the space.
+
+#### Storage vars
+
+- **voting_delay** = 600   
+The time in seconds to wait between when a proposal is created and when the proposal voting period starts. 
+
+- **voting_period** = 3600   
+The duration of a proposal in seconds.
+
+- **proposal_threshold** = 1   
+The minimum amount of voting power that is required to submit a proposal.
+
+- **voting_strategy** = 0x123...   
+The voting strategy contract address that is used to calculate the voting power for each voter, the strategy will return 1 voting power for any address. 
+
+- **auth** = 0x123...  
+The authenticator contract address that is used to verify the addresses of voters and proposers.
+
+- **next_proposal_nonce** = 1  
+An increasing nonce to make the proposal id unique.
+
+- **proposals** = TBD  
+A mapping of proposals by ids with required information: execution_hash, start, end.
+
+- **votes** = TBD  
+A mapping of votes by proposal id with required information: address, choice, and voting power.
+
+#### Functions
+
+- **vote(address, proposal, choice)**  
+Function to cast a vote, the payload is made up of the Ethereum address of the voter, the proposal id and the choice.
+
+- **propose(address, execution_hash, metadata_uri)**   
+Submit a new proposal. The **address** is an Ethereum address, **execution_hash** is a hash from EIP-712 that include execution details, here is an example https://gist.github.com/bonustrack/45fdb2f0235d6ad3ccffa328234379aa the **metadata_uri** is an URL string that leads to a JSON file which includes proposal metadata like title or body, this is similar to NFT tokenURI. It can be "ipfs://AbC..." or "https://..." URL.
+
```

Comment:
I think we need another function. Since the space contract is called by the `Authenticator`, and `Authenticator` only has a `execute` function, which cannot know if it needs to call `vote` or `propose`.
So we either need to have an entry point here that will then call `propose` or `vote`, or put it in the `authenticator`. Thoughts @bonustrack @Orland0x ?

Reply:
The function execute need to know what method to call, this info can come from the param **data** or we must come with another way to do it. I used the same naming as DSProxy from MakerDAO where "data" are calldata see https://github.com/makerdao/dss-proxy/blob/master/src/DssProxy.sol

Reply:
Ok. So one way or the other, `data` needs to hold `function_selector` (as well as `sig, execution_hash, and metadata_uri`)